src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/AnalyzerWorkQueue.cs

```diff
@@ -12,98 +12,118 @@ namespace OmniSharp.Roslyn.CSharp.Workers.Diagnostics
 {
     public class AnalyzerWorkQueue
     {
-        private readonly int _throttlingMs = 300;
+        private class Queue
+        {
+            public Queue(TimeSpan throttling)
+            {
+                Throttling = throttling;
+            }
 
-        private readonly ConcurrentDictionary<DocumentId, (DateTime modified, CancellationTokenSource workDoneSource)> _workQueue =
-            new ConcurrentDictionary<DocumentId, (DateTime modified, CancellationTokenSource workDoneSource)>();
+            public ImmutableHashSet<DocumentId> WorkWaitingToExecute { get; set; } = ImmutableHashSet<DocumentId>.Empty;
+            public ImmutableHashSet<DocumentId> WorkExecuting { get; set; } = ImmutableHashSet<DocumentId>.Empty;
+            public DateTime LastThrottlingBegan { get; set; } = DateTime.UtcNow;
+            public TimeSpan Throttling { get; }
+            public CancellationTokenSource WorkPendingToken { get; set; }
+        }
 
-        private readonly ConcurrentDictionary<DocumentId, (DateTime modified,  CancellationTokenSource workDoneSource)> _currentWork =
-            new ConcurrentDictionary<DocumentId, (DateTime modified, CancellationTokenSource workDoneSource)>();
+        private readonly Dictionary<AnalyzerWorkType, Queue> _queues = null;
 
+        private readonly ILogger<AnalyzerWorkQueue> _logger;
         private readonly Func<DateTime> _utcNow;
         private readonly int _maximumDelayWhenWaitingForResults;
-        private readonly ILogger<AnalyzerWorkQueue> _logger;
+        private readonly object _queueLock = new Object();
 
-        public AnalyzerWorkQueue(ILoggerFactory loggerFactory, Func<DateTime> utcNow = null, int timeoutForPendingWorkMs = 15*1000)
+        public AnalyzerWorkQueue(ILoggerFactory loggerFactory, Func<DateTime> utcNow = null, int timeoutForPendingWorkMs = 15 * 1000)
```

Comment:
should this timeout be configurable in omnisharp.json?

Reply:
In previous version i think it could, but now in this version it only applies to request for diagnostics from single file. It's basically just failsafe to prevent api hang in this version. However i am not sure is this needed any more at all since theres timeout for file analysis too on worker side which should prevent that situation.

This timeout could be valid to remove altogether. Timeout for single file analysis in worker could indeed be configurable and that way it's implemented only on single location.